[nnc] Added a new API to distribute loop and all its parents

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61293**
* #61087
* #61085

Differential Revision: [D29560008](https://our.internmc.facebook.com/intern/diff/D29560008)

navahgar committed:
[nnc] Added a new API to distribute loop and all its parents

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29560008).

ZolotukhinM commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -5260,6 +5260,118 @@ TEST(LoopNest, DistributeLoopOverInnerLoops) {
+# CHECK-NOT: for (
+        )IR";
+
+    auto newForI = dynamic_cast<For*>(Stmt::clone(forI));
+    auto forM = For::make(m, 0, 50, newForI);
+    auto par = Block::make({forM});
+    LoopNest nest(par, {a_buf.node(), b_buf.node()});
+    auto new_loops = LoopNest::distributeLoopAndParents(newForI);
```
Nit: a combination of camelCase and snake_case :)

navahgar replied:
Fixed.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -238,6 +238,9 @@ class TORCH_API LoopNest {
   // S6:    for k
   // S7:      B[i] = B[i] +
   static std::vector<For*> distributeLoop(For* loop);
+  // Same as above, but also distribute parent loops.
+  // Returns the result of distributing the outermost loop.
+  static std::vector<For*> distributeLoopAndParents(For* loop);
```
Could you please add an example of what it does? I have hard times understanding what exactly "also distribute parent loops" means.

navahgar replied:
Added an example for these 2 APIs as well.

ZolotukhinM reviewed (approved):
Overall looks good, I have only minor nits.

navahgar committed:
Update on "[nnc] Added a new API to distribute loop and all its parents"


Differential Revision: [D29560008](https://our.internmc.facebook.com/intern/diff/D29560008)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29560008).

navahgar committed:
Update on "[nnc] Added a new API to distribute loop and all its parents"


Differential Revision: [D29560008](https://our.internmc.facebook.com/intern/diff/D29560008)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29560008).
